### DISCUSSION TITLE
What is the sweetspot for trained images per subject?

### DISCUSSION THREAD

**Discussion by nicnet123** [exadel-inc/CompreFace#1053] [General]
Aim: create recognized face metadata for my collection of photos.
Here's my current workflow (bash scripting):

100k photos with one or more faces in each one
1. use detection and imagemagick to create set of images with single face in each ("single-face-images")
2. train recognition with some (perhaps 10) of my most common people
3. use recognition (e.g. similarity=0.9) to move single-face-images to folders (folder name = person name from recognition)
4. move any single-face-image not recognized to directory in the form ./face_unknown_folder/GENDER/AGE
5. manual checking of faces in the face_unknown folders, use these to do some more recognition trainings - at least the gender and age does some sorting for me
6. then go back to step 2 and train more faces, and repeat over and over, until I can't be bothered to do more (i.e. I have trainings for my most important subjects)
7. use recognition (with large subject count, based on the previous iterations) on original multiface photos, and write face recognitions into meta-data.

Question - as I iterate through steps 2 to 6, I'm building large collection of named single-face-images.
Should I be using ALL of these as trainings for recognition?
What is the sweetspot for trained images per subject? (1/10/100/1000/100000?)
With this method, I could quickly create thousands of images per subject, which of course in principle would increase accuracy. 
But it seems to me a bit silly to have a training for EVERY instance of a subject - and I suspect will impact performance.

I've tested photoprism - really good face recognition, but currently not enough flexibility in writing metadata, and digikam - too much effort to get many faces reliably recognized.
Your offering seems to fix both those issues for me.

**Comment by Insktall** [comment]
I would like to answer this question,The sweet spot for trained images per subject can vary depending on several factors such as the complexity of the subject, the diversity of the images, and the quality of the dataset. However, as a general rule of thumb, a minimum of 100-200 images per subject is recommended for satisfactory performance of many image recognition models.Having more images per subject can potentially improve the accuracy and robustness of the model, but it can also increase the time and resources required for training. It is important to strike a balance between the number of images and the training time/resources.
Additionally, it's important to ensure that the dataset is diverse enough to cover the variations in the subject, including different lighting conditions, angles, and backgrounds. A diverse and well-curated dataset can improve the model's ability to generalize and perform well on new and unseen images.  https://tinyurl.com/9sfeffmk

**Comment by pospielov** [comment]
I'll try to give an intuitive explanation.
Imagine you are working as security, given three photos of different people, and your goal is to allow or deny entry, depending on who the person is. One important detail - you've never seen them before, so you really need to compare the photo and the person.
1. Is one photo enough?
One photo is enough, in case it has good quality and it's recent. 
2. Will more photos help?
If you have more photos of the person, it will probably increase the chance you will recognize him. However, there is a bigger chance you will make a mistake and think to a stranger that this is a person from a photo. 
So more photos increase the accuracy for the person but also increase false positives. 
3. What will help?
It's better to choose one photo but with as good quality as possible. 
Also, when you make a picture to recognize, it should also be as good quality as possible. 
4. Is it possible to make use of more photos?
I think yes, but it's not a trivial logic.
E.g., imagine you have not one, but 10 photos of each person.
And the person you are trying to recognize is very similar to one of the 10 photos of person A, but not very similar to the other 9 photos of person A. On the other hand, the person is similar to all 10 photos of person B, but not so much as for this one photo of person A. 
What would you decide to do? 
In different situations, you will make different decisions:
1. In some situations, you will decide that this is person A, as one of his photos is the best match.
2. In some situations, you will decide that there is a bigger chance that this is a person B
3. In some situations, you will decide it is too dangerous to allow him to enter, so you'll deny it.
